src/voices.ts

```diff
@@ -0,0 +1,491 @@
+
+import { novelty, quality, recommended, veryLowQuality, TGender, TQuality, IRecommended, defaultRegion } from "./data.js";
+
+// export type TOS = 'Android' | 'ChromeOS' | 'iOS' | 'iPadOS' | 'macOS' | 'Windows';
+// export type TBrowser = 'ChromeDesktop' | 'Edge' | 'Firefox' | 'Safari';
+
+const navigatorLanguages = () => window?.navigator?.languages || [];
+const navigatorLang = () => (navigator?.language || "").split("-")[0].toLowerCase();
+
+export interface IVoices {
+    label: string;
+    voiceURI: string;
+    name: string;
+    language: string;
+    gender?: TGender | undefined;
+    age?: string | undefined;
+    offlineAvailability: boolean;
+    quality?: TQuality | undefined;
+    pitchControl: boolean;
+    recommendedPitch?: number | undefined;
+    recommendedRate?: number | undefined;
+}
+
+const normalQuality = Object.values(quality).map(({ normal }) => normal);
+const highQuality = Object.values(quality).map(({ high }) => high);
+
+function compareQuality(a?: TQuality, b?: TQuality): number {
+    const qualityToNumber = (quality: TQuality) => {
+        switch (quality) {
+            case "veryLow": {return 0;}
+            case "low": {return 1;}
+            case "normal": {return 2;}
+            case "high": {return 3;}
+            case "veryHigh": {return 4;}
+            default: {return -1};
+        }
+    }
+
+    return qualityToNumber(b || "low") - qualityToNumber(a || "low");
+};
+
+export async function getSpeechSynthesisVoices(): Promise<SpeechSynthesisVoice[]> {
+    const a = () => speechSynthesis.getVoices();
+    
+    const voices = a();
+    if (Array.isArray(voices) && voices.length) return voices;
+
+    return new Promise((resolve, _reject) => {
+
+        let counter = 1000;
+        const tick = () => {
+            if (counter < 1) return resolve([]);
+            // console.log(counter);
+            
+            --counter;
+            const voices = a();
+            if (Array.isArray(voices) && voices.length) return resolve(voices);
+            setTimeout(tick, 10);
+        }
+        setTimeout(tick, 10);
```

Comment:
`setTimeout()` has a resolution of about 20ms, last time I checked. Also, IIRC this is affected by browser window visibility (it doesn't really matter for this use-case though)

Reply:
... I think that a 200ms tick would be fine for this.